Neolution.Extensions.Identity/TokenSignManager.cs

```diff
@@ -0,0 +1,272 @@
+                    return null;
+                }
+
+                // Continue with the rest of your sign-in logic
+                this.logger.LogInformation("OIDC token is valid for user with email={Email}", email);
+
+                // TODO: Think about different discovery options for external users
+                var user = await this.userManager.FindByEmailAsync(email);
+                if (user is null)
+                {
+                    return null;
+                }
+
+                var error = await this.signInManager.PreSignInCheckAsync(user);
+                if (error != null)
+                {
+                    return null;
+                }
+
+                this.logger.LogWarning("OIDC token sign-in for user with id={UserId} failed", user.Id);
+                return user;
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex, "Could not sign in user with OIDC token");
+                return null;
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task<JsonWebToken?> CreateAccessTokenAsync(TUser user, string? authenticationMethod)
+        {
+            this.logger.LogInformation("Create access token for user with id={UserId}", user.Id);
+            var claims = await this.userManager.GetClaimsAsync(user);
+            if (authenticationMethod != null)
+            {
+                this.logger.LogInformation("Set authentication method to {AuthenticationMethod}", authenticationMethod);
+                claims.Add(new Claim(ClaimTypes.AuthenticationMethod, authenticationMethod));
+            }
+
+            return this.jwtGenerator.GenerateAccessToken(user, claims);
+        }
+
+        /// <summary>
+        /// Resets the lockout of the specified user.
+        /// </summary>
+        /// <param name="user">The user.</param>
+        /// <returns>The identity result.</returns>
+        private async Task<IdentityResult> ResetLockoutWithResultAsync(TUser user)
+        {
+            if (!this.userManager.SupportsUserLockout)
+            {
+                this.logger.LogTrace("User lockout is disabled, did not reset lockout for user id={UserId}", user.Id);
+                return IdentityResult.Success;
+            }
+
+            return await this.userManager.ResetAccessFailedCountAsync(user) ?? IdentityResult.Success;
+        }
+
+        /// <summary>
+        /// Validates the OpenID connect token. Throws an exception if the token is invalid.
+        /// </summary>
+        /// <param name="token">The token.</param>
+        /// <returns>The <see cref="ClaimsPrincipal"/>.</returns>
+        private async Task<ClaimsPrincipal> ValidateOpenIdConnectInfoAsync(OpenIdConnectToken token)
+        {
+            var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(token.DiscoveryDocumentUrl, new OpenIdConnectConfigurationRetriever(), new HttpDocumentRetriever());
```

Comment:
@neoscie  is this going to be cached or downloaded every time?

Reply:
There is some sort of cache. I'm not sure about the defaults and how it works exactly. Here is the options class to configure it: https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.configuration.lastknowngoodconfigurationcacheoptions?view=msal-web-dotnet-latest